src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                CancellationToken cancellationToken)
+            {
+                _session = session;
+                _data = data;
+                _recentItemsManager = recentItemsManager;
+                _globalOptions = globalOptions;
+                _cancellationToken = cancellationToken;
+
+                _filterText = _session.ApplicableToSpan.GetText(_data.Snapshot);
+                _initialRoslynTriggerKind = Helpers.GetRoslynTriggerKind(_data.InitialTrigger);
+
+                if (!_session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+                {
+                    // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                    // For now, the default hasSuggestedItemOptions is false.
+                    hasSuggestedItemOptions = false;
+                }
+
+                _hasSuggestedItemOptions = hasSuggestedItemOptions || _data.DisplaySuggestionItem;
+
+                // We prefer using the original snapshot, which should always be available from items provided by Roslyn's CompletionSource.
+                // Only use data.Snapshot in the theoretically possible but rare case when all items we are handling are from some non-Roslyn CompletionSource.
+                var snapshotForDocument = TryGetInitialTriggerLocation(_data, out var intialTriggerLocation)
+                    ? intialTriggerLocation.Snapshot
+                    : _data.Snapshot;
+
+                _document = snapshotForDocument?.TextBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                _completionService = _document?.GetLanguageService<CompletionService>();
+                _completionRules = _completionService?.GetRules(CompletionOptions.From(_document!.Project)) ?? CompletionRules.Default;
+
+                // Let us make the completion Helper used for non-Roslyn items case-sensitive.
+                // We can change this if get requests from partner teams.
+                _completionHelper = _document != null ? CompletionHelper.GetHelper(_document) : new CompletionHelper(isCaseSensitive: true);
+
+                // Nothing to highlight if user hasn't typed anything yet.
+                _highlightMatchingPortions = _filterText.Length > 0
+                    && _globalOptions.GetOption(CompletionViewOptions.HighlightMatchingPortionsOfCompletionListItems, _document?.Project.Language);
+
+            }
+
+            public FilteredCompletionModel? UpdateCompletionList()
+            {
+                if (ShouldDismissCompletionListImmediately())
+                    return null;
+
+                // Determine the list of items to be included in the completion list.
+                // This is computed based on the filter text as well as the current
+                // selection of filters and expander.
+                using var _ = ComputeItemsToBeIncluded(out var itemsToBeIncluded);
```

Comment:
it may just be me, but i really dont' like nested methods passing out disposers.  I would just make the items list here.  and then rename the method to be AddCompletionListItems.